### ISSUE TITLE
Fearless concurrency: how Clojure Rust Pony Erlang and Dart let you achieve that. - Renato Athaydes

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2783]
Fearless concurrency: how Clojure, Rust, Pony, Erlang and Dart let you achieve that. - Renato Athaydes<br>
<br>
https://ift.tt/2tzJrjH<br>
<br>
<br>
<br>
<div>
<div dir="ltr">
<p>Anyone who has written any concurrent code (i.e. code where more than one Thread of execution is present) involving more than a couple of mutable variables knows that it's really hard to get it right using the low-level tools like locks and semaphores.</p>
<p><br></p>
<div>
<p>Testing this kind of code is even harder: you're forced to write probabilistic tests (i.e. tests that can prove the presence of a bug, but never its absence with 100% certainty even after a large number of runs) or use</p>
<a rel="nofollow" href="http://www.cs.umd.edu/projects/PL/multithreadedtc/overview.html" target="_blank">tools</a>
<p>specifically designed to test concurrency (which can be pretty complex themselves).</p>
</div>
<p><br></p>
<div>
<p>Most reasonable developers tend to use</p>
<a rel="nofollow" href="https://docs.oracle.com/javase/tutorial/essential/concurrency/highlevel.html" target="_blank">higher level concurrency primitives</a>
<p>(thread-local, concurrent collections etc) to write concurrent code if at all possible. The problem is, that sometimes, unfortunately, these tools are just not sufficient, it's still easy to shoot your own foot and get lost in a sea of complexity.</p>
</div>
<p><br></p>
<p>For these reasons, several models that make it easier to reason about concurrent programs have been envisioned over time.</p>
<p><br></p>
<p>In this article, we'll have a quick look at a few of them, from new to not-so-new languages. I don't intend to give an extensive analysis of each solution, or make a formal comparison between them. My intention is to simply explain the basics of each solution and how they can be used in practice (with code samples that show off what the result of using the models might look like), so that other developers may have an easier time understanding them and deciding which solution, or language, might be better applicable to their particular problems.</p>
<p><br></p>
<p><br></p>
<p><span><i><font size="3">Concurrency in this article is meant to refer to the multi-threaded execution of tasks over possibly multiple CPUs, or even machines.</font></i></span></p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<h2>
<a name="TOC-Alternative-concurrency-models" id="TOC-Alternative-concurrency-models"></a>Alternative concurrency models</h2>
</div>
<p><br></p>
</div>
<div>
<h3>
<a name="TOC-Immutability-Functional-Programming-Clojure-" id="TOC-Immutability-Functional-Programming-Clojure-"></a>Immutability, Functional Programming (Clojure)<br>
</h3>
</div>
<p><br></p>
<p>A popular alternative to make concurrent code safe and easier to understand is to just remove mutable variables entirely from the equation.</p>
<p><br></p>
<p>After all, when all things are immutable, concurrency becomes much more manageable, as functional programming languages have shown for a long time. Shared state is not a problem unless it's <b>mutable</b> state.<br></p>
<p><br></p>
<p>It may appear to be impossible to write an application using only immutable data structures in case you're not familiar with functional programming, but that's definitely not the case. To understand how, let's have a look at a very simple example.<br></p>
<p><br></p>
<div>
<p>The following example shows how</p>
<a rel="nofollow" href="https://clojure.org/" target="_blank">Clojure</a>
<p>allows replacing some element of an immutable list (or vector):</p>
</div>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span></p>
<p><span>&nbsp;(</span><span>def</span> <span>my-vec [</span><span>0</span> <span>1</span> <span>2</span> <span>3</span> <span>4</span> <span>5</span><span>])</span></p>
<p><span>; replace element at index 3 with x</span></p>
<p><span>(</span><span>def</span> <span>new-vec (</span><span>assoc</span> <span>my-vec</span> <span>3</span> <span>:x))</span></p>
<p><span>(</span><span>println</span> <span>my-vec)</span></p>
<p><span>(</span><span>println</span> <span>new-vec)</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<p><code>[0 1 2 3 4 5]</code><br>
<br>
<code>[0 1 2 :x 4 5]</code></p>
<p><br></p>
<p><br></p>
<p>The original vector, <code>my-vec</code>, is not modified. It's immutable, after all... but we can create another vector, <code>new-vec</code>, which contains the contents of the original vector, but with one or more elements replaced, as in this example where the 4th element is replaced with <code>:x</code>.</p>
<p><br></p>
<p>Programs that are written using only immutable data structure always have to work this way: instead of changing something, you apply a function to it to get a new reference to the "modified" something. For other parts of the code to "see" changes, they must explicitly ask for the new reference, as the copy they might have had will never change.<br></p>
<p><br></p>
<p>Such programs can be trivially parallelized because there's never any need to synchronize access to immutable state (and there's no mutable state!).</p>
<p><br></p>
<p>The example below shows how we can transform each element of a vector by running some expensive calculation on a different Thread for each item (explanation below):</p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span>(</span><span>defn</span> <span>random-long-between</span></p>
<p><span>[min max]</span></p>
<p><span>(</span><span>def</span> <span>max-rand (</span><span>-</span> <span>max min))</span></p>
<p><span>(</span><span>long</span> <span>(</span><span>+</span> <span>min (</span><span>rand</span> <span>max-rand))))</span></p>
<p><span>(</span><span>defn</span> <span>expensive-operation</span></p>
<p><span>[n]</span></p>
<p><span>(</span><span>let</span> <span>[wait-millis (</span><span>random-long-between</span> <span>250</span> <span>750</span><span>)]</span></p>
<p><span>(</span><span>Thread/sleep</span> <span>wait-millis)</span></p>
<p><span>{:n n :even? (</span><span>even?</span> <span>n) :delay wait-millis}))</span></p>
<p><span>(</span><span>defn</span> <span>eager-map-async</span></p>
<p><span>"Applies fn fun on each element of the given sequence on a separate Thread,</span></p>
<p><span>returning a vector with the results."</span></p>
<p><span>[fun sequence]</span></p>
<p><span>; future applies the given function asynchronously in another Thread</span></p>
<p><span>(</span><span>let</span> <span>[async-seq (</span><span>map</span> <span>(</span><span>fn</span> <span>[n] (</span><span>future</span> <span>(</span><span>fun</span> <span>n))) sequence)]</span></p>
<p><span>; we then use '@' to de-reference each future, getting the new elements into a vector</span></p>
<p><span>(</span><span>vec</span> <span>(</span><span>map</span> <span>(</span><span>fn</span> <span>[fut] @fut) async-seq))))</span></p>
</div>
<p><span><br></span></p>
<div>
<span>&nbsp;(</span><span>def</span> <span>my-vec [</span><span>0</span> <span>1</span> <span>2</span> <span>3</span> <span>4</span> <span>5</span><span>])</span>
<p><span>; Apply and time the eager-map-async function on my-vec</span></p>
<p><span>(</span><span>def</span> <span>new-vec (</span><span>time</span> <span>(</span><span>eager-map-async</span> <span>expensive-operation my-vec)))</span></p>
<p><span>(</span><span>doall</span> <span>(</span><span>map</span> <span>println new-vec))</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<div>
<p><code><br>
"Elapsed time: 622.28397 msecs"<br></code><br>
<code>{:n 0, :even? true, :delay 616}<br></code><br>
<code>{:n 1, :even? false, :delay 557}<br></code><br>
<code>{:n 2, :even? true, :delay 254}<br></code><br>
<code>{:n 3, :even? false, :delay 611}<br></code><br>
<code>{:n 4, :even? true, :delay 584}<br></code><br>
<code>{:n 5, :even? false, :delay 257}<br>
<br></code></p>
</div>
<p><br></p>
<p><br></p>
<div>
<p>The</p>
<code>eager-map-async</code>
<p>function in the code above does the main job of calling the provided function,</p>
<code>expensive-operation</code>
<p>in this case, in a different Thread by using the</p>
<a rel="nofollow" href="https://clojuredocs.org/clojure.core/future" target="_blank"><code>future</code></a>
<p>function, part of the Clojure SDK. It then blocks waiting for all elements to be realized by</p>
<code>future</code>
<p>(using the @ symbol to de-reference the future), before returning a vector containing all the results.</p>
</div>
<p><br></p>
<p>You can see that the full vector transformation took <b>622 ms</b> to complete, which is expected if all operations ran in parallel, as the longest delay was <b>616 ms</b> (presumably, the other 6 ms were spent doing actual work).<br></p>
<p><br></p>
<p><br></p>
<div><span><font size="3"><i>This example only touches upon Clojure's concurrency facilities! You can learn more about concurrency in Clojure on the <a rel="nofollow" href="https://www.braveclojure.com/concurrency/" target="_blank">Brave Clojure</a> online book.</i></font></span></div>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<p>It may sound very wasteful to make copies all the time, but it's possible to avoid fully copying data structures on each modification via structural sharing and other techniques. Clojure's</p>
<a rel="nofollow" href="https://clojure.org/reference/data_structures" target="_blank">data structures</a>
<p>are very smart in that regard!</p>
<a rel="nofollow" href="https://hypirion.com/musings/understanding-persistent-vector-pt-1" target="_blank">This article</a>
<p>on</p>
<i>hypirion.com</i>
<p>explains in detail how they work. If you just want to learn to use them,</p>
<a rel="nofollow" href="https://purelyfunctional.tv/guide/clojure-collections/" target="_blank">purelyfunctional.tv</a>
<p>has an ultimate guide on Clojure collections.</p>
</div>
<p><br></p>
<p>But immutability and functional programming are not the only game in town. There are other ways to write concurrent applications with guarantees about correctness (even though it might still be hard to write them!).</p>
<p><br></p>
<p>The next section briefly explains the Rust solution to the problem.<br></p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<p>One of the most intriguing solutions to the concurrency problem is the system used by Rust: the</p>
<a rel="nofollow" href="https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html" target="_blank">ownership system</a>
<p>, supported by the (in)famous</p>
<a rel="nofollow" href="http://www.squidarth.com/rc/rust/2018/05/31/rust-borrowing-and-ownership.html" target="_blank">Rust borrow checker</a>
<p>. With this system, you can write concurrent code as you always did in languages like Java or C++, with the difference that any mistake you make will be caught by the borrow checker. If there's a possibility that you passed a mutable variable to a function, and that variable may be accessed simultaneously from another Thread, the Rust borrow checker will yell at you and your code won't even compile, let alone run. It's great, but the cost is high: programs that can be easily written in most languages can be taxing to write in Rust, even when they would be safe to run without all the protections that Rust demands.</p>
</div>
<p><br></p>
<p><br></p>
<div><font size="3"><span><i>The ownership system was devised primarily as a solution to memory management without a garbage-collector. However, the system also solves concurrency issues, with a little help of the type system, because memory management and concurrency are actually <a rel="nofollow" href="https://doc.rust-lang.org/book/ch16-00-concurrency.html" target="_blank">intrinsically linked</a>.</i></span></font></div>
<p><br></p>
<p><br></p>
<p>The following example demonstrates how Rust handles mutable variables (even in the absence of concurrency, as in Rust, anything is visible from all Threads, so the possibility of concurrent access is always there):<br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span>use</span> <span>std</span><span>::</span><span>collections</span><span>::</span><span>LinkedList;</span></p>
<p><span>fn</span> <span>main</span><span>() {</span></p>
<p><span>let</span> <span>mut</span> <span>list: LinkedList</span><span>&lt;</span><span>u64</span><span>&gt;</span> <span>=</span> <span>LinkedList</span><span>::</span><span>new</span><span>();</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>1</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>2</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>3</span><span>);</span></p>
<p><span>println!</span><span>(</span><span>"{:?}"</span><span>,</span> <span>&amp;</span><span>list);</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p>Notice that the declaration of <code>list</code> includes the mutable keyword, <code>mut</code>. Failing to include that keyword would make the list immutable, so the <code>push_back</code> method could not be used in that case:<br></p>
<p><br></p>
<p><br></p>
<div>
<p><code>error[E0596]: cannot borrow `list` as mutable, as it is not declared as mutable</code><br>
<code>&nbsp;--&gt; src/main.rs:9:5</code><br>
<code>&nbsp; |</code><br>
<code>8 |&nbsp;&nbsp;&nbsp;&nbsp; let list: LinkedList&lt;u64&gt; = LinkedList::new();</code><br>
<code>&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ---- help: consider changing this to be mutable: `mut list`</code><br>
<code>9 |&nbsp;&nbsp;&nbsp;&nbsp; list.push_back(1);</code><br>
<code>&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp; ^^^^ cannot borrow as mutable</code><br></p>
</div>
<p><br></p>
<p>This shows the borrow checker in action. When you call a method in Rust, the receiver, <code>list</code> in this case, needs to be <i>borrowed</i> by the method implementation itself! When you borrow something that you might want to modify, you must borrow it as mutable.</p>
<p><br></p>
<p>This is what the <code>push_back</code> implementation does:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
pub</span> <span>fn</span> <span>push_back</span><span>(</span><span>&amp;</span><span>mut</span> <span>self</span><span>, elt: T) { ... }<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>This allows Rust to know when a value may be changed by a block of code, which helps it reason about what's safe to allow.</p>
<p><br></p>
<p>Let's say we wanted to add the elements to the list in a separate function:<br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span>use</span> <span>std</span><span>::</span><span>collections</span><span>::</span><span>LinkedList;</span></p>
<p><span>fn</span> <span>main</span><span>() {</span></p>
<p><span>let</span> <span>mut</span> <span>list: LinkedList</span><span>&lt;</span><span>u64</span><span>&gt;</span> <span>=</span> <span>LinkedList</span><span>::</span><span>new</span><span>();</span></p>
<p><span>add_elements_to</span><span>(</span><span>&amp;</span><span>mut</span> <span>list);</span></p>
<p><span>println!</span><span>(</span><span>"{:?}"</span><span>,</span> <span>&amp;</span><span>list);</span></p>
<p><span>}</span></p>
<p><span>fn</span> <span>add_elements_to</span><span>(list:</span> <span>&amp;</span><span>mut</span> <span>LinkedList&lt;</span><span>u64</span><span>&gt;) {</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>1</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>2</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>3</span><span>);</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:<br></p>
<p><br></p>
<p>That still works and is safe, so Rust allows it. But suppose we wanted to do this in a separate thread... now, that's no longer safe, so Rust won't let you do that:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span>use</span> <span>std</span><span>::</span><span>collections</span><span>::</span><span>LinkedList;</span></p>
<p><span>fn</span> <span>main</span><span>() {</span></p>
<p><span>let</span> <span>mut</span> <span>list: LinkedList</span><span>&lt;</span><span>u64</span><span>&gt;</span> <span>=</span> <span>LinkedList</span><span>::</span><span>new</span><span>();</span></p>
<p><span>thread</span><span>::</span><span>spawn</span><span>(</span><span>move</span> <span>||</span> <span>{</span></p>
<p><span>add_elements_to</span><span>(</span><span>&amp;</span><span>mut</span> <span>list);</span></p>
<p><span>}).</span><span>join</span><span>().</span><span>unwrap</span><span>();</span></p>
<p><span>println!</span><span>(</span><span>"{:?}"</span><span>,</span> <span>&amp;</span><span>list);</span> <span>// ERROR!!!!</span></p>
<p><span>}</span></p>
<p><span>fn</span> <span>add_elements_to</span><span>(list:</span> <span>&amp;</span><span>mut</span> <span>LinkedList&lt;</span><span>u64</span><span>&gt;) {</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>1</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>2</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>3</span><span>);</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Error:</p>
<p><br></p>
<div>
<p><br></p>
<p><code>&nbsp; --&gt; src/main.rs:12:22</code><br>
<code>&nbsp;&nbsp; |</code><br>
<code>9&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp; thread::spawn(move || {</code><br>
<code>&nbsp;&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ------- value moved into closure here</code><br>
<code>10 |&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; add_elements_to(&amp;mut list);</code><br>
<code>&nbsp;&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ---- variable moved due to use in closure</code><br>
<code>11 |&nbsp;&nbsp;&nbsp;&nbsp; }).join().unwrap();</code><br>
<code>12 |&nbsp;&nbsp;&nbsp;&nbsp; println!("{:?}", &amp;list);</code><br>
<code>&nbsp;&nbsp; |&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ^^^^^ value borrowed here after move</code><br>
<code>&nbsp;&nbsp; |</code><br>
<code>&nbsp;&nbsp; = note: move occurs because `list` has type `std::collections::LinkedList&lt;u64&gt;`, which does not implement the `Copy` trait</code></p>
</div>
<p><br></p>
<p>Error messages in Rust are extremely nice in most cases. You can clearly see the Rust developers put a lot of effort into making things clear for users when errors occur.</p>
<p><br></p>
<p>In our case above, it clearly informs us that we can't use <code>list</code> after we moved it (notice the <code>move</code> keyword when we declared the Thread's closure, which implicitly moves any variables the closure uses into the new scope) into the closure's scope as it does not implement the <code>Copy</code> trait (which presumably would have allowed that). And we had to move it because it's not thead-safe to allow two different threads to have access to a mutable variable.<br></p>
<p><br></p>
<div>
<p>Rust has a nice solution to this problem:</p>
<a rel="nofollow" href="https://doc.rust-lang.org/book/ch16-02-message-passing.html" target="_blank">channels</a>
<p>.</p>
</div>
<p><br></p>
<p>Below, we use a channel to be able to get back our linked-list after giving it to another thread:<br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
use</span> <span>std</span><span>::</span><span>thread;</span></p>
<p><span>use</span> <span>std</span><span>::</span><span>collections</span><span>::</span><span>LinkedList;</span></p>
<p><span>use</span> <span>std</span><span>::</span><span>sync</span><span>::</span><span>mpsc;</span></p>
<p><span>fn</span> <span>main</span><span>() {</span></p>
<p><span>let</span> <span>(tx, rx)</span> <span>=</span> <span>mpsc</span><span>::</span><span>channel</span><span>();</span></p>
<p><span>let</span> <span>mut</span> <span>list: LinkedList</span><span>&lt;</span><span>u64</span><span>&gt;</span> <span>=</span> <span>LinkedList</span><span>::</span><span>new</span><span>();</span></p>
<p><span>thread</span><span>::</span><span>spawn</span><span>(</span><span>move</span> <span>||</span> <span>{</span></p>
<p><span>add_elements_to</span><span>(</span><span>&amp;</span><span>mut</span> <span>list);</span></p>
<p><span>tx.</span><span>send</span><span>(list).</span><span>unwrap</span><span>();</span></p>
<p><span>}).</span><span>join</span><span>().</span><span>unwrap</span><span>();</span></p>
<p><span>let</span> <span>list</span> <span>=</span> <span>rx.</span><span>recv</span><span>().</span><span>unwrap</span><span>();</span></p>
<p><span>println!</span><span>(</span><span>"{:?}"</span><span>,</span> <span>&amp;</span><span>list);</span></p>
<p><span>}</span></p>
<p><span>fn</span> <span>add_elements_to</span><span>(list:</span> <span>&amp;</span><span>mut</span> <span>LinkedList&lt;</span><span>u64</span><span>&gt;) {</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>1</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>2</span><span>);</span></p>
<p><span>list.</span><span>push_back</span><span>(</span><span>3</span><span>);</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<p>This style of handling concurrency safely without locks is called <i>message passing</i>, and is one of the core concepts of the Actor model, which we'll encounter in the next sections. But Rust also lets you use lower-level concurrency tools.<br></p>
<p><br></p>
<div>
<p>Let's look at the example the</p>
<a rel="nofollow" href="https://doc.rust-lang.org/book/ch16-03-shared-state.html" target="_blank">Rust Book</a>
<p>gives on how to use a</p>
<code>Mutex</code>
<p>to handle shared mutable state (in this case, a counter):</p>
</div>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
use</span> <span>std</span><span>::</span><span>sync</span><span>::</span><span>{Mutex, Arc};</span></p>
<p><span>use</span> <span>std</span><span>::</span><span>thread;</span></p>
<p><span>fn</span> <span>main</span><span>() {</span></p>
<p><span>let</span> <span>counter</span> <span>=</span> <span>Arc</span><span>::</span><span>new</span><span>(Mutex</span><span>::</span><span>new</span><span>(</span><span>0</span><span>));</span></p>
<p><span>let</span> <span>mut</span> <span>handles</span> <span>=</span> <span>vec!</span><span>[];</span></p>
<p><span>for</span> <span>_</span> <span>in</span> <span>0</span><span>..</span><span>10</span> <span>{</span></p>
<p><span>let</span> <span>counter</span> <span>=</span> <span>Arc</span><span>::</span><span>clone</span><span>(</span><span>&amp;</span><span>counter);</span></p>
<p><span>let</span> <span>handle</span> <span>=</span> <span>thread</span><span>::</span><span>spawn</span><span>(</span><span>move</span> <span>||</span> <span>{</span></p>
<p><span>let</span> <span>mut</span> <span>num</span> <span>=</span> <span>counter.</span><span>lock</span><span>().</span><span>unwrap</span><span>();</span></p>
<p><span>*</span><span>num</span> <span>+=</span> <span>1</span><span>;</span></p>
<p><span>});</span></p>
<p><span>handles.</span><span>push</span><span>(handle);</span></p>
<p><span>}</span></p>
<p><span>for</span> <span>handle</span> <span>in</span> <span>handles {</span></p>
<p><span>handle.</span><span>join</span><span>().</span><span>unwrap</span><span>();</span></p>
<p><span>}</span></p>
<p><span>println!</span><span>(</span><span>"Result: {}"</span><span>,</span> <span>*</span><span>counter.</span><span>lock</span><span>().</span><span>unwrap</span><span>());</span></p>
<p><span>}<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:<br></p>
<p><br></p>
<p><br></p>
<p>This example starts up 10 Threads at the same time (approximately) and then waits for all of them to run. Each Thread accesses and changes the counter, which at the end is printed by the <code>main</code> function. The reason Rust allows that is that before each Thread accesses the counter, they obtain a lock that protects it from concurrent access (the lock is automatically released when the <code>Arc</code> it's in goes out of scope).<br></p>
<p><code><br></code></p>
<p><code><br></code></p>
<p><font size="3"><span><i><font size="4"><code>Arc</code></font> stands for atomic reference count and is the secret for sharing state between multiple threads in Rust.</i></span></font></p>
<p><br></p>
<p><br></p>
<p>This is good old-fashioned concurrency with locks, but with a little help from a smarter compiler!<br></p>
<p><br></p>
<div>
<p>Next, we'll look at another system language, but one that</p>
<a rel="nofollow" href="https://www.ponylang.io/discover/#what-is-pony" target="_blank">guarantees</a>
<p>the absence of deadlocks and race conditions.</p>
</div>
<p><br></p>
<div>
<h3>
<a name="TOC-Reference-Capabilities-Pony-" id="TOC-Reference-Capabilities-Pony-"></a>Reference Capabilities (Pony)</h3>
</div>
<p><br></p>
<p><br></p>
<div>
<p>Another interesting alternative is being popularized by the</p>
<a rel="nofollow" href="https://www.ponylang.io/" target="_blank">Pony programming language</a>
<p>: it uses a</p>
<a rel="nofollow" href="https://tutorial.ponylang.io/capabilities.html" target="_blank">capabilities system</a>
<p>to indicate which variables are safe to share between threads (or</p>
<a rel="nofollow" href="https://tutorial.ponylang.io/types/actors.html" target="_blank"><i>Actors</i></a>
<p>, which are similar to</p>
<a rel="nofollow" href="https://en.wikipedia.org/wiki/Green_threads" target="_blank">green threads</a>
<p>in Pony). For example, if a variable is immutable (a</p>
<i>val</i>
<p>in Pony) then Pony will let you share it with other Actors without any limitations. If the variable is</p>
<i>iso</i>
<p>(for isolated), meaning there's only one reference to it at any given time (which the compiler verifies), then it's also safe to share it even if it's mutable (because no two Actors can write to it at the same time). And there are a few other capabilities that give you fine control to define how your variables should be used, with the compiler ensuring it's safe to use them as such.</p>
</div>
<p><br></p>
<p>To get a taste of what that looks like, let's write a small Pony program that passes a mutable variable, say, an <code>Array[U64]</code>, between two Actors. The first actor, <code>Main</code>, creates an array, sends it to the <code>Adder</code> actor which knows how to add a number <code>n</code> to each element of the array, then sends the result back to <code>Main</code>, which prints it.<br></p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><br></p>
<p><span>actor</span> <span>Main</span></p>
<p><span>let</span> <span>_env:</span> <span>Env</span></p>
<p><span>let</span> <span>_adder:</span> <span>Adder</span></p>
<p><span>new</span> <span>create</span><span>(env:</span> <span>Env</span><span>) =&gt;</span></p>
<p><span>_env = env</span></p>
<p><span>_adder =</span> <span>Adder</span><span>(this)</span></p>
<p><span>start</span><span>()</span></p>
<p><span>be</span> <span>start</span><span>() =&gt;</span></p>
<p><span>let</span> <span>my_array:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>iso</span> <span>= [</span><span>1</span><span>;</span> <span>2</span><span>;</span> <span>3</span><span>;</span> <span>4</span><span>]</span></p>
<p><span>_adder.</span><span>addN</span><span>(</span><span>consume</span> <span>my_array,</span> <span>4</span><span>)</span></p>
<p><span>be</span> <span>showResult</span><span>(result:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>iso</span><span>) =&gt;</span></p>
<p><span>let</span> <span>a:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>box</span> <span>=</span> <span>consume</span> <span>result</span></p>
<p><span>let</span> <span>len = a.</span><span>size</span><span>() -</span> <span>1</span></p>
<p><span>let</span> <span>str =</span> <span>recover</span> <span>String</span> <span>end</span></p>
<p><span>str.</span><span>append</span><span>(</span><span>"["</span><span>)</span></p>
<p><span>for</span> <span>(i, item)</span> <span>in</span> <span>a.</span><span>pairs</span><span>()</span> <span>do</span></p>
<p><span>str.</span><span>append</span><span>(item.</span><span>string</span><span>())</span></p>
<p><span>if</span> <span>i &lt; len</span> <span>then</span></p>
<p><span>str.</span><span>append</span><span>(</span><span>", "</span><span>)</span></p>
<p><span>end</span></p>
<p><span>end</span></p>
<p><span>str.</span><span>append</span><span>(</span><span>"]"</span><span>)</span></p>
<p><span>_env.</span><span>out</span><span>.</span><span>print</span><span>(</span><span>consume</span> <span>str)</span></p>
<p><span>actor</span> <span>Adder</span></p>
<p><span>let</span> <span>_runner:</span> <span>Main</span></p>
<p><span>new</span> <span>create</span><span>(runner:</span> <span>Main</span><span>) =&gt;</span></p>
<p><span>_runner = runner</span></p>
<p><span>be</span> <span>addN</span><span>(array:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>iso</span><span>, n:</span> <span>U64</span><span>) =&gt;</span></p>
<p><span>let</span> <span>result =</span> <span>recover</span> <span>iso</span></p>
<p><span>let</span> <span>a:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>ref</span> <span>=</span> <span>consume</span> <span>array</span></p>
<p><span>for</span> <span>(i, item)</span> <span>in</span> <span>a.</span><span>pairs</span><span>()</span> <span>do</span></p>
<p><span>try</span> <span>a.</span><span>update</span><span>(i, item + n)?</span> <span>end</span></p>
<p><span>end</span></p>
<p><span>a</span></p>
<p><span>end</span></p>
<p><span>_runner.</span><span>showResult</span><span>(</span><span>consume</span> <span>result)</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p>Notice that the array in question is always the same array, no copy is made of the original. But as we're using Pony Actors, the array is definitely being passed between threads (assuming the two actors are run in different threads, which may not be always the case), but unlike in most languages, this is completely safe (memory- and concurrency-safe) in Pony.</p>
<p><br></p>
<p>To understand why, a more detailed explanation is due.</p>
<p><br></p>
<div>
<p>First of all, notice that the</p>
<i>methods</i>
<p>of the actors in the example above are declared with the</p>
<code>be</code>
<p>keyword. That means they are not actually methods in the usual sense of the word (which Pony does have, they are declared with</p>
<code>fun</code>
<p>), they are</p>
<b>behaviours</b>
<p>.</p>
<a rel="nofollow" href="https://tutorial.ponylang.io/types/actors.html#behaviours" target="_blank">Behaviours</a>
<p>are executed asynchronously (and concurrently with other behaviours!). For that reason, the arguments they take must be</p>
<i>sendable</i>
<p>, and they never return any value directly.</p>
</div>
<p><br></p>
<p>As we saw earlier, it's possible to send a <b>mutable</b> variable to another Actor if and only if it's an <i>iso</i> variable (it's also possible to send <i>val</i>s as they are immutable, and <i>tags</i> which are basically other Actors). That's why we declare <code>addN</code> (which wants to mutate the array it's given) like this:</p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>be</span> <span>addN</span><span>(array:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>iso</span><span>, n:</span> <span>U64</span><span>) =&gt; ...<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>As you can see, the capability is declared after the type. <code>Array[U64] iso</code> means an array of unsigned-64-bit integers with the <i>iso</i> capability. When calling this behaviour, the owner of the array must give it up, which is done with <code>consume</code>:<br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
_adder.</span><span>addN</span><span>(</span><span>consume</span> <span>my_array,</span> <span>4</span><span>)<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p><br></p>
<p>The <code>n</code> argument is a Primitive, so it's a <i>val</i> and can always be passed safely to other Actors. By consuming the <code>my_array</code> variable, we ensure that there's still only one reference to it, and that reference now belongs to the body of the <code>addN</code> behaviour.<br></p>
<p><br></p>
<p><br></p>
<div><span><font size="3"><i>Pony <a rel="nofollow" href="https://tutorial.ponylang.io/types/primitives.html" target="_blank">Primitives</a>, interestingly, are different from other languages in that they can be user-defined. Combined with union types, they allow for very efficient and concise code.</i></font></span></div>
<p><br></p>
<p><br></p>
<p>Looking at the implementation of the <code>addN</code> behaviour, notice the following line:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>let</span> <span>a:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>ref</span> <span>=</span> <span>consume</span> <span>array<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<div>
<p>This may look pointless, but it's necessary because</p>
<code>array</code>
<p>here has the</p>
<i>iso</i>
<p>capability, but to modify the array's internals we need the</p>
<code>ref</code>
<p>capability (see</p>
<a rel="nofollow" href="https://stdlib.ponylang.io/builtin-Array/#update" target="_blank">update</a>
<p>method). The</p>
<i>ref</i>
<p>capability allows us to read and write to a variable. So, the above line basically casts an</p>
<i>iso</i>
<p>to a</p>
<i>ref</i>
<p>. This is safe as long as we don't pass the</p>
<i>ref</i>
<p>around, which we are not allowed to do by the compiler.</p>
</div>
<p><br></p>
<p>Finally, it's interesting to notice that by the end of the <code>addN</code> behaviour, we'll need to pass the array back to the Main actor, but we can't do that yet as our array now is a <i>ref</i> (and we consumed the original <code>array</code> reference, so it can't be used anymore). In order to <i>recover</i> the <i>iso</i> capability that will let us do that, we wrap the whole part of the code where we mutate the array inside a <code>recover</code> block.</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>let</span> <span>result =</span> <span>recover</span> <span>iso</span></p>
<p><span>let</span> <span>a:</span> <span>Array</span><span>[</span><span>U64</span><span>]</span> <span>ref</span> <span>=</span> <span>consume</span> <span>array</span></p>
<p><span>for</span> <span>(i, item)</span> <span>in</span> <span>a.</span><span>pairs</span><span>()</span> <span>do</span></p>
<p><span>try</span> <span>a.</span><span>update</span><span>(i, item + n)?</span> <span>end</span></p>
<p><span>end</span></p>
<p><span>a</span></p>
<p><span>end</span></p>
<p><span>_runner.</span><span>showResult</span><span>(</span><span>consume</span> <span>result)<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<div>
<p>The Pony docs</p>
<a rel="nofollow" href="https://tutorial.ponylang.io/reference-capabilities/recovering-capabilities.html" target="_blank">explain this</a>
<p>in more detail, but the important thing to understand is that within a</p>
<code>recover</code>
<p>block, only</p>
<i>sendable</i>
<p>values from the enclosing lexical scope can be accessed, hence the result of the block is also</p>
<i>sendable</i>
<p>. So we end up with a</p>
<i>sendable</i>
<p>result!</p>
</div>
<p><br></p>
<p><br></p>
<div>
<p>It's absolutely brilliant, but unfortunately, as far as I know, Pony has not been gaining much popularity in the several years it's already been around... I've been following its progress for a while as a curious bystander, and it's a bit sad to see that, even though Pony brings something quite innovative to the table, it looks like that's just not enough to attract many people to it. That may be due to the tough competition, with Rust taking most of the interest from people working on high performance, highly concurrent systems. But I suspect it's due to the lack of tooling (I even wrote a</p>
<a rel="nofollow" href="https://github.com/renatoathaydes/pony-gradle-plugin" target="_blank">Gradle plugin</a>
<p>for it and tried to contribute to the VS Code plugin at some point but was met with total lack of interest from the people running the project, so gave up) and the lack of a larger community around it, which is always a problem for any new language.</p>
</div>
<p><br></p>
<p>But I still have some hope for Pony, as I think people may come to realize that provably safe concurrency (no deadlocks, no race conditions) and high performance (advanced, per-actor garbage-collector with predictable pauses) at the same time is kind of a big deal.<br></p>
<p><br></p>
<div>
<h3>
<a name="TOC-The-Actor-Model-Erlang-Dart-" id="TOC-The-Actor-Model-Erlang-Dart-"></a>The Actor Model (Erlang, Dart)</h3>
</div>
<p><br></p>
<p><br></p>
<div>
<p>The last alternative solution to writing concurrent applications safely that I am aware of is the</p>
<a rel="nofollow" href="https://en.wikipedia.org/wiki/Actor_model" target="_blank">Actor Model</a>
<p>. It had already been invented by the 1970's, as most of the greatest ideas in computer science, but it's still slowly making its way into mainstream.</p>
</div>
<p><br></p>
<div>
<p>Some languages that are based on the Actor Model include</p>
<a rel="nofollow" href="https://www.erlang.org/" target="_blank">Erlang</a>
<p>and its sister language,</p>
<a rel="nofollow" href="https://elixir-lang.org/" target="_blank">Elixir</a>
<p>. Other languages have libraries to support the Actor Model, for example,</p>
<a rel="nofollow" href="https://akka.io/" target="_blank">Akka</a>
<p>for Java/Scala,</p>
<a rel="nofollow" href="https://riker.rs/" target="_blank">Riker</a>
<p>for Rust,</p>
<a rel="nofollow" href="http://www.actor-framework.org/" target="_blank">CAF</a>
<p>for C++. There's even a cross-platform library called</p>
<a rel="nofollow" href="http://proto.actor/" target="_blank">Proto.Actor</a>
<p>which enables Go, .Net and Java/Kotlin actors to communicate with each other.</p>
</div>
<p><br></p>
<p><br></p>
<p><span><font size="3"><i>Pony is also part of this category, but due to the quite different way messages work in Pony (for example, they can be mutable) and its unique capabilities system, I decided to keep it in its own special category.</i></font></span></p>
<p><br></p>
<p><br></p>
<p>In the Actor Model, each Actor has its own state, which no other Actor can interact with except via message passing. Hence, concurrency is achieved transparently by the runtime, with the actual application code never having to worry about that.</p>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
-</span><span>module</span><span>(</span><span>tut15</span><span>).</span></p>
<p><span>-</span><span>export</span><span>([</span><span>start</span><span>/</span><span>0</span><span>,</span> <span>ping</span><span>/</span><span>2</span><span>,</span> <span>pong</span><span>/</span><span>0</span><span>]).</span></p>
<p><span>ping</span><span>(</span><span>0</span><span>,</span> <span>Pong_PID</span><span>)</span> <span>-&gt;</span></p>
<p><span>Pong_PID</span> <span>!</span> <span>finished,</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"ping finished~n"</span><span>, []);</span></p>
<p><span>ping</span><span>(</span><span>N</span><span>,</span> <span>Pong_PID</span><span>)</span> <span>-&gt;</span></p>
<p><span>Pong_PID</span> <span>!</span> <span>{ping,</span> <span>self</span><span>()},</span></p>
<p><span>receive</span></p>
<p><span>pong -&gt;</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"Ping received pong~n"</span><span>, [])</span></p>
<p><span>end</span><span>,</span></p>
<p><span>ping</span><span>(</span><span>N</span> <span>-</span> <span>1</span><span>,</span> <span>Pong_PID</span><span>).</span></p>
<p><span>pong</span><span>()</span> <span>-&gt;</span></p>
<p><span>receive</span></p>
<p><span>finished -&gt;</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"Pong finished~n"</span><span>, []);</span></p>
<p><span>{ping,</span> <span>Ping_PID</span><span>} -&gt;</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"Pong received ping~n"</span><span>, []),</span></p>
<p><span>Ping_PID</span> <span>!</span> <span>pong,</span></p>
<p><span>pong</span><span>()</span></p>
<p><span>end</span><span>.</span></p>
<p><span>start</span><span>()</span> <span>-&gt;</span></p>
<p><span>Pong_PID</span> <span>=</span> <span>spawn</span><span>(tut15, pong, []),</span></p>
<p><span>spawn</span><span>(tut15, ping, [</span><span>3</span><span>,</span> <span>Pong_PID</span><span>]).<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<span><code>1&gt;</code></span> <span><code>c(tut15).</code></span><code><br>
{ok,tut15}</code>
<pre>
<span><code>2&gt; </code></span><span><code>tut15: start().</code></span><code>
&lt;</code><span><code>0.36</code></span><code>.</code><span><code>0</code></span><code>&gt;
Pong received ping
Ping received pong
Pong received ping
Ping received pong
Pong received ping
Ping received pong
ping finished
Pong finished</code>
</pre>
</div>
</div>
<p><br></p>
<p>Erlang is a traditional functional language, so as you can see, actors are really just functions that receive/send messages to each other. The syntax to send a message to an actor, which Erlang calls process (not an OS process, Erlang processes are much cheaper to create and many thousands of them can exist at any given time) is:</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p>A process PID can be obtained by spawning a new actor, which is done by calling <code>spawn</code> with the module and function to run, as well as the initial message to give it:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>Pong_PID</span> <span>=</span> <span>spawn</span><span>(tut15, pong, [])<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>The ping function then can send a message to the process with this PID:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>Pong_PID</span> <span>!</span> <span>{ping,</span> <span>self</span><span>()}<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>On the receiver side, pattern-matching is used to handle the message (as Erlang is dynamically typed, patterns match on contents rather than types, unlike with Pony and, as we'll see, Dart) within a <code>receive</code> block:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
pong</span><span>()</span> <span>-&gt;</span></p>
<p><span>receive</span></p>
<p><span>finished -&gt;</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"Pong finished~n"</span><span>, []);</span></p>
<p><span>{ping,</span> <span>Ping_PID</span><span>} -&gt;</span></p>
<p><span>io</span><span>:</span><span>format</span><span>(</span><span>"Pong received ping~n"</span><span>, []),</span></p>
<p><span>Ping_PID</span> <span>!</span> <span>pong,</span></p>
<p><span>pong</span><span>()</span></p>
<p><span>end</span><span>.</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p><br></p>
<p>Very neat.</p>
<p><br></p>
<p><br></p>
<p>Here's the roughly equivalent ping-pong example in Dart:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
import</span> <span>'dart:io'</span> <span>show</span> <span>exit;</span></p>
<p><span>import</span> <span>'dart:isolate'</span><span>;</span></p>
<p><span>class</span> <span>PingStartMessage</span> <span>{</span></p>
<p><span>final</span> <span>int</span> <span>n;</span></p>
<p><span>final</span> <span>SendPort</span> <span>pong;</span></p>
<p><span>PingStartMessage</span><span>(</span><span>this</span><span>.n,</span> <span>this</span><span>.pong);</span></p>
<p><span>}</span></p>
<p><span>ping</span><span>(</span><span>PingStartMessage</span> <span>startMessage) {</span></p>
<p><span>var</span> <span>n</span> <span>=</span> <span>startMessage.n;</span></p>
<p><span>final</span> <span>pong</span> <span>=</span> <span>startMessage.pong;</span></p>
<p><span>final</span> <span>pingPort</span> <span>=</span> <span>ReceivePort</span><span>();</span></p>
<p><span>pingPort.</span><span>listen</span><span>((message) {</span></p>
<p><span>if</span> <span>(message</span> <span>==</span> <span>'pong'</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"Ping received pong"</span><span>);</span></p>
<p><span>if</span> <span>(n</span> <span>&gt;</span> <span>0</span><span>) {</span></p>
<p><span>pong.</span><span>send</span><span>(pingPort.sendPort);</span></p>
<p><span>n</span><span>--</span><span>;</span></p>
<p><span>}</span> <span>else</span> <span>{</span></p>
<p><span>pong.</span><span>send</span><span>(</span><span>'finished'</span><span>);</span></p>
<p><span>}</span></p>
<p><span>}</span></p>
<p><span>});</span></p>
<p><span>pong.</span><span>send</span><span>(pingPort.sendPort);</span></p>
<p><span>}</span></p>
<p><span>pong</span><span>(</span><span>SendPort</span> <span>starterPort) {</span></p>
<p><span>final</span> <span>pongPort</span> <span>=</span> <span>ReceivePort</span><span>();</span></p>
<p><span>pongPort.</span><span>listen</span><span>((message) {</span></p>
<p><span>if</span> <span>(message</span> <span>==</span> <span>'finished'</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"Pong finished"</span><span>);</span></p>
<p><span>starterPort.</span><span>send</span><span>(message);</span></p>
<p><span>}</span> <span>else</span> <span>if</span> <span>(message</span> <span>is</span> <span>SendPort</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"Pong received ping"</span><span>);</span></p>
<p><span>message.</span><span>send</span><span>(</span><span>'pong'</span><span>);</span></p>
<p><span>}</span></p>
<p><span>});</span></p>
<p><span>starterPort.</span><span>send</span><span>(pongPort.sendPort);</span></p>
<p><span>}</span></p>
<p><span>main</span><span>()</span> <span>async</span> <span>{</span></p>
<p><span>final</span> <span>starterPort</span> <span>=</span> <span>ReceivePort</span><span>();</span></p>
<p><span>await</span> <span>Isolate</span><span>.</span><span>spawn</span><span>(pong, starterPort.sendPort);</span></p>
<p><span>starterPort.</span><span>listen</span><span>((message) {</span></p>
<p><span>if</span> <span>(message</span> <span>is</span> <span>SendPort</span><span>) {</span></p>
<p><span>// got pong's port, give it to ping</span></p>
<p><span>Isolate</span><span>.</span><span>spawn</span><span>(ping,</span> <span>PingStartMessage</span><span>(</span><span>3</span><span>, message));</span></p>
<p><span>}</span> <span>else</span> <span>if</span> <span>(message</span> <span>==</span> <span>'finished'</span><span>) {</span></p>
<p><span>exit</span><span>(</span><span>0</span><span>);</span></p>
<p><span>}</span></p>
<p><span>});</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p>Result:</p>
<p><br></p>
<p><br></p>
<p><code>Pong received ping</code><br>
<code>Ping received pong</code><br>
<code>Pong received ping</code><br>
<code>Ping received pong</code><br>
<code>Pong received ping</code><br>
<code>Ping received pong</code><br>
<code>Pong received ping</code><br>
<code>Ping received pong</code><br>
<code>Pong finished</code></p>
<p><br></p>
<p><br></p>
<div>
<p>It's a little bit more verbose than Erlang as we need to do some more maintenance to start the Isolates and manage their ports ourselves (see the</p>
<a href="https://pub.dartlang.org/packages/isolate" target="_blank">isolate</a>
<p>library, which helps with the boilerplate). But the essence is the same.</p>
</div>
<p><br></p>
<div>
<p>Without Isolates, Dart is actually a</p>
<a href="https://www.dartlang.org/tutorials/language/futures" target="_blank">single-threaded language</a>
<p>, even considering its async/await feature. Asynchronous functions are simply run in what Dart calls</p>
<a href="https://webdev.dartlang.org/articles/performance/event-loop" target="_blank">microtasks</a>
<p>, which is its event loop, similarly to JavaScript.</p>
</div>
<p><br></p>
<p><br></p>
<p><span><i><font size="3">Notice that all Dart code runs in an Isolate, including main. But if the application code never starts any other Isolates, the whole application runs in a single Thread, on a single CPU.</font></i></span></p>
<div><span><i><font size="3">As of Dart 2, Isolates are not supported in web applications - they need to use <a href="https://api.dartlang.org/stable/2.1.1/dart-html/Worker-class.html" target="_blank">Web Workers</a> instead. However, they work fine in <a href="https://flutter.dev/" target="_blank">Flutter</a>.</font></i></span></div>
<p><br></p>
<p><br></p>
<p>Isolates, however, run in parallel with other Isolates. They do not share any memory with other Isolates, hence are very similar to Erlang processes.</p>
<p><br></p>
<p>In the above example, we can see that several different types of messages were being exchanged. For example, when the ping isolate was started, we gave it a <code>PingStartMessage</code> right away:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>Isolate</span><span>.</span><span>spawn</span><span>(ping,</span> <span>PingStartMessage</span><span>(</span><span>3</span><span>, message));<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>The pong Isolate responds to a <code>SendPort</code> message (the type of the object used to send messages to other Isolates) with the <code>pong</code> string:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>if</span> <span>(message</span> <span>is</span> <span>SendPort</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"Pong received ping"</span><span>);</span></p>
<p><span>message.</span><span>send</span><span>(</span><span>'pong'</span><span>);</span></p>
<p><span>}<br>
<br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>When the ping Isolate receives that, it might send another ping or a <code>finished</code> message:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br></span><span>if</span> <span>(message</span> <span>==</span> <span>'pong'</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"Ping received pong"</span><span>);</span></p>
<p><span>if</span> <span>(n</span> <span>&gt;</span> <span>0</span><span>) {</span></p>
<p><span>pong.</span><span>send</span><span>(pingPort.sendPort);</span></p>
<p><span>n</span><span>--</span><span>;</span></p>
<p><span>}</span> <span>else</span> <span>{</span></p>
<p><span>pong.</span><span>send</span><span>(</span><span>'finished'</span><span>);</span></p>
<p><span>}</span></p>
<p><span>}</span></p>
<p><span><br></span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>All messages being sent in this example are immutable, but Dart does not actually require that, which can be confusing.<br></p>
<p><br></p>
<p>This example demonstrates that:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
import</span> <span>'dart:isolate'</span><span>;</span></p>
<p><span>import</span> <span>'dart:io'</span> <span>show</span> <span>exit;</span></p>
<p><span>class</span> <span>Message</span> <span>{</span></p>
<p><span>final</span> <span>List</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>list;</span></p>
<p><span>final</span> <span>SendPort</span> <span>sender;</span></p>
<p><span>Message</span><span>(</span><span>this</span><span>.list,</span> <span>this</span><span>.sender);</span></p>
<p><span>}</span></p>
<p><span>go</span><span>(</span><span>Message</span> <span>msg) {</span></p>
<p><span>msg.list.</span><span>add</span><span>(</span><span>10</span><span>);</span></p>
<p><span>print</span><span>(</span><span>"Added 10 to the list: ${</span><span>msg.list</span><span>}"</span><span>);</span></p>
<p><span>msg.sender.</span><span>send</span><span>(</span><span>'ok'</span><span>);</span></p>
<p><span>}</span></p>
<p><span>main</span><span>()</span> <span>async</span> <span>{</span></p>
<p><span>final</span> <span>port</span> <span>=</span> <span>ReceivePort</span><span>();</span></p>
<p><span>final</span> <span>list</span> <span>=</span> <span>[</span><span>1</span><span>,</span> <span>2</span><span>,</span> <span>3</span><span>];</span></p>
<p><span>await</span> <span>Isolate</span><span>.</span><span>spawn</span><span>(go,</span> <span>Message</span><span>(list, port.sendPort));</span></p>
<p><span>port.</span><span>listen</span><span>((msg) {</span></p>
<p><span>if</span> <span>(msg</span> <span>==</span> <span>'ok'</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"List after receiving message: $</span><span>list</span><span>"</span><span>);</span></p>
<p><span>exit</span><span>(</span><span>0</span><span>);</span></p>
<p><span>}</span></p>
<p><span>});</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<p><br></p>
<div>
<p><code>Added 10 to the list: [1, 2, 3, 10]</code><br>
<code>List after receiving message: [1, 2, 3]</code></p>
</div>
<p>The reason is simple: again, no memory is shared between Dart Isolates. The List the Isolate in the example above modifies is NOT the same list the main function created. Messages are copied when they cross the Isolate boundary. For that reason, for something like the above to work, it's necessary that the Isolate send back the result of its computation. Otherwise, there's no way for any Isolates to observe the changes made by it.</p>
<p><br></p>
<p>Given this knowledge, we can rewrite the example above correctly now:</p>
<p><br></p>
<p><br></p>
<div>
<div>
<p><span><br>
import</span> <span>'dart:isolate'</span><span>;</span></p>
<p><span>import</span> <span>'dart:io'</span> <span>show</span> <span>exit;</span></p>
<p><span>class</span> <span>Message</span> <span>{</span></p>
<p><span>final</span> <span>List</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>list;</span></p>
<p><span>final</span> <span>SendPort</span> <span>sender;</span></p>
<p><span>Message</span><span>(</span><span>this</span><span>.list,</span> <span>this</span><span>.sender);</span></p>
<p><span>}</span></p>
<p><span>go</span><span>(</span><span>Message</span> <span>msg) {</span></p>
<p><span>msg.list.</span><span>add</span><span>(</span><span>10</span><span>);</span></p>
<p><span>print</span><span>(</span><span>"Added 10 to the list: ${</span><span>msg.list</span><span>}"</span><span>);</span></p>
<p><span>msg.sender.</span><span>send</span><span>(msg.list);</span></p>
<p><span>}</span></p>
<p><span>main</span><span>()</span> <span>async</span> <span>{</span></p>
<p><span>final</span> <span>port</span> <span>=</span> <span>ReceivePort</span><span>();</span></p>
<p><span>final</span> <span>list</span> <span>=</span> <span>[</span><span>1</span><span>,</span> <span>2</span><span>,</span> <span>3</span><span>];</span></p>
<p><span>await</span> <span>Isolate</span><span>.</span><span>spawn</span><span>(go,</span> <span>Message</span><span>(list, port.sendPort));</span></p>
<p><span>port.</span><span>listen</span><span>((msg) {</span></p>
<p><span>if</span> <span>(msg</span> <span>is</span> <span>List</span><span>&lt;</span><span>int</span><span>&gt;</span><span>) {</span></p>
<p><span>print</span><span>(</span><span>"List after receiving message: $</span><span>msg</span><span>"</span><span>);</span></p>
<p><span>exit</span><span>(</span><span>0</span><span>);</span></p>
<p><span>}</span></p>
<p><span>});</span></p>
<p><span>}</span></p>
</div>
</div>
<p><br></p>
<p><br></p>
<p>Result:</p>
<div>
<p><code>Added 10 to the list: [1, 2, 3, 10]</code><br>
<code>List after receiving message: [1, 2, 3, 10]</code></p>
</div>
<p><br></p>
<p><br></p>
<p><br></p>
<div>
<div>
<p>Notice that this article does not include Go, a language that admittedly has an elegant concurrency solution as well (</p>
<a rel="nofollow" href="https://gobyexample.com/channels" target="_blank">Go channels</a>
<p>) because that solution is</p>
<a href="https://blog.golang.org/race-detector" target="_blank">not actually thread-safe</a>
<p>- it's not very hard to have race conditions in Go or corrupt state because Go does not enforce a separation of shareable and not-shareable mutable state.</p>
</div>
<p><br></p>
<div>
<p>For the same reason, I decided to not include</p>
<a rel="nofollow" href="https://kotlinlang.org/docs/reference/coroutines/basics.html" target="_blank">Kotlin co-routines</a>
<p>either. They suffer from essentially the same problem.</p>
</div>
<p><br></p>
<p>It should be clear to the reader how the solutions presented here differ from that.</p>
<p><br></p>
<div>
<p>I need to also mention</p>
<a rel="nofollow" href="http://book.realworldhaskell.org/read/concurrent-and-multicore-programming.html" target="_blank">Haskell</a>
<p>would probably be a better example than Clojure of a functional language that suppoorts easy concurrency because of the central roles immutable data plays in it. I only decided to use Clojure because of my familiarity with it and the fact that it seems quite a lot easier to manage concurrency in Clojure from my biased point-of-view. Hope the Haskell fans out there will forgive me :).</p>
</div>
</div>
<p><br></p>
<p><br></p>
<p><br></p>
<p>This was a long article, but I hope that you've learned something new about concurrency that will be useful to you regardless of which language you use.</p>
<p><br></p>
<p>The different solutions discussed in this article do not form clean, separate categories, but I believe the grouping made sense. The important for me was to try to capture the essence of each model, and show how they help&nbsp; make sure concurrency is done safely.</p>
<p><br></p>
<p>In Clojure, even though mutability is actually supported via some lower-level primitives and the JVM itself (all Clojure code can call any Java code), as long as users stick to the immutable collections in concurrent code (which is very easy to do, and even idiomatic in Clojure) they will be completely safe from concurrency issues.</p>
<p><br></p>
<p>It is true that it's possible for any language to use this strategy (including Java and C++), but functional programming languages in general tend to be much better in dealing with immutable data - it's their sweet spot.<br></p>
<p><br></p>
<p>Rust, on the other hand, at the cost of requiring a detailed description by the programmer of what can be changed, borrowed, sent to other threads etc. even in code that is not intended to be used concurrently, allows the programmer to achieve a high level of confidence in the thread-safety of their programs, and with very little runtime costs.<br></p>
<p><br></p>
<p>Pony is mid-way between Rust and Erlang. It has extremely high performance, but generally looks and feels like a much higher level language. It has such a sexy syntax (looks like Python, but no, it doesn't use syntactically meaningful whitespace!) that it makes it hard for me to not want to write all my code in it! But the lack of libraries and community, as well as tooling, really hurts its adoption.<br></p>
<p><br></p>
<p>Erlang, the oldest and most battle-tested language in the list, offers a superbly reliable VM and it (or Elixir) should definitely be considered by anyone wishing to write a highly available, distributed system that doesn't necessarily require very high bare metal performance. Its use of dynamic typing makes me a little bit hesitant to use it, though, as I really love the help provided by static typing.<br></p>
<p><br></p>
<p>Dart has a very approachable concurrent system that anyone familiar with Java, for example, can learn in a matter of minutes. The async/await support is amazing, and Isolates are fairly easy to code against.</p>
<p>It is important to understand, though, that Dart may let you write code that doesn't behave correctly without warning if you don't understand the limitations of Isolates (complete isolation of memory). This limitation is not something that may be lifted in the future with more work from the Dart team: it's the fundamental feature of Dart that allows it, as Erlang, to support real parallelism without any concurrency primitive at all.</p>
<p><br></p>
<p>Finally, I'd like to mention that there's no best system, in my opinion. For each occasion, any of the systems above could be the best solution. They are all very good, but have different strengths and weaknesses which I hope the reader will be better equipped to discern after reading this article and doing a little bit further research based on the information provided.</p>
<p><br></p>
<p><br></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Sign in - Google Accounts<br>
<br>
February 26, 2019 at 11:48AM
